src/navigable-group/internal.tsx

```diff
@@ -0,0 +1,168 @@
+
+const InternalNavigableGroup = forwardRef(
+  (
+    {
+      children,
+      getItemKey,
+      navigationDirection = 'horizontal',
+      __internalRootRef,
+      ...props
+    }: InternalNavigableGroupProps,
+    ref: React.Ref<NavigableGroupProps.Ref>
+  ) => {
+    const baseProps = getBaseProps(props);
+    const navigationAPI = useRef<SingleTabStopNavigationAPI>(null);
+    const containerObjectRef = useRef<HTMLDivElement>(null);
+    const containerRef = useMergeRefs(containerObjectRef, __internalRootRef);
+    const focusedIdRef = useRef<string>();
+
+    useImperativeHandle(ref, () => ({
+      focus: () => {
+        const target = getNextFocusTarget();
+        if (target) {
+          target.focus();
+        }
+      },
+    }));
+
+    function getNextFocusTarget(): null | HTMLElement {
+      if (containerObjectRef.current) {
+        const focusables = getFocusablesFrom(containerObjectRef.current);
+        return focusables.find(button => getItemKey(button) === focusedIdRef.current) ?? focusables[0] ?? null;
+      }
+      return null;
+    }
+
+    function onUnregisterActive(focusableElement: HTMLElement) {
+      // Only refocus when the node is actually removed (no such element anymore).
+      const target = navigationAPI.current?.getFocusTarget();
+
+      if (target && getItemKey(target) !== getItemKey(focusableElement)) {
+        target.focus();
+      }
+    }
+
+    useEffect(() => {
+      navigationAPI.current?.updateFocusTarget();
+    });
+
+    function onFocus(event: React.FocusEvent) {
+      if (event.target instanceof HTMLElement && navigationAPI.current?.isRegistered(event.target)) {
+        focusedIdRef.current = getItemKey(event.target);
+      }
+      navigationAPI.current?.updateFocusTarget();
+    }
+
+    function onBlur() {
+      navigationAPI.current?.updateFocusTarget();
+    }
+
+    function onKeyDown(event: React.KeyboardEvent) {
+      const focusTarget = navigationAPI.current?.getFocusTarget();
+      let specialKeys = [];
+      switch (navigationDirection) {
+        case 'horizontal':
+          specialKeys = [KeyCode.right, KeyCode.left];
+          break;
+        case 'vertical':
+          specialKeys = [KeyCode.down, KeyCode.up];
+          break;
+        case 'both':
+          specialKeys = [KeyCode.right, KeyCode.left, KeyCode.down, KeyCode.up];
+          break;
+      }
+      specialKeys.push(KeyCode.end, KeyCode.home, KeyCode.pageUp, KeyCode.pageDown);
+      if (hasModifierKeys(event) || specialKeys.indexOf(event.keyCode) === -1) {
+        return;
+      }
+      if (!containerObjectRef.current || !focusTarget) {
+        return;
+      }
+      // Ignore navigation when the focused element is not a registered focusable.
+      if (document.activeElement && !navigationAPI.current?.isRegistered(document.activeElement)) {
+        return;
+      }
+      event.preventDefault();
+
+      const focusables = getFocusablesFrom(containerObjectRef.current);
```

Comment:
Do we know that `focusables` will be always an non-empty array? I'm wondering if there should add an additional early return in case of it's empty. E.g. I've elements in the navigable group got disabled while navigating.